token/core/zkatdlog/crypto/rp/bulletproof.go

```diff
@@ -0,0 +1,422 @@
+		// these are randomly generated arrays
+		randomLeft = append(randomLeft, p.Curve.NewRandomZr(rand))
+		randomRight = append(randomRight, p.Curve.NewRandomZr(rand))
+	}
+
+	// C commits to L = (b_0, ..., b_{p.BitLength}) and R = (b_0 - 1 , ..., b_{p.BitLength} - 1)
+	C := commitVector(left, right, p.LeftGenerators, p.RightGenerators, p.Curve)
+	// C is a hiding commitment thanks to rho
+	C.Add(p.P.Mul(rho))
+
+	// D commits two random vectors U and V
+	D := commitVector(randomLeft, randomRight, p.LeftGenerators, p.RightGenerators, p.Curve)
+	// D is a hiding commitment thanks to eta
+	D.Add(p.P.Mul(eta))
+
+	array := common.GetG1Array([]*math.G1{C, D})
+	// compute challenges y and z
+	bytesToHash, err := array.Bytes()
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+	y := p.Curve.HashToZr(bytesToHash)
+	z := p.Curve.HashToZr(y.Bytes())
+
+	leftPrime := make([]*math.Zr, len(left))
+	rightPrime := make([]*math.Zr, len(right))
+
+	randRightPrime := make([]*math.Zr, len(randomRight))
+
+	zPrime := make([]*math.Zr, len(left))
+	// z^2
+	zSquare := z.PowMod(p.Curve.NewZrFromInt(2))
+
+	for i := 0; i < len(left); i++ {
+		// compute L_i - z
+		leftPrime[i] = p.Curve.ModSub(left[i], z, p.Curve.GroupOrder)
+		// compute R_i + z
+		rightPrime[i] = p.Curve.ModAdd(right[i], z, p.Curve.GroupOrder)
+		// compute y^i
+		y2i := y.PowMod(p.Curve.NewZrFromInt(int64(i)))
+		// compute (R_i+z)y^i
+		rightPrime[i] = p.Curve.ModMul(rightPrime[i], y2i, p.Curve.GroupOrder)
+		// compute V_iy^i
+		randRightPrime[i] = p.Curve.ModMul(randomRight[i], y2i, p.Curve.GroupOrder)
+		// compute 2^iz^2
+		zPrime[i] = p.Curve.ModMul(zSquare, p.Curve.NewZrFromInt(2).PowMod(p.Curve.NewZrFromInt(int64(i))), p.Curve.GroupOrder)
+	}
+
+	// compute \sum y^iV_i(L_i-z)
+	t1 := innerProduct(leftPrime, randRightPrime, p.Curve)
+	// compute \sum y^i(V_i(L_i-z) + (R_i +z)U_i)
+	t1 = p.Curve.ModAdd(t1, innerProduct(rightPrime, randomLeft, p.Curve), p.Curve.GroupOrder)
+	// compute \sum y^i(V_i(L_i-z) + (R_i+z)U_i) + U_i2^iz^2
+	t1 = p.Curve.ModAdd(t1, innerProduct(zPrime, randomLeft, p.Curve), p.Curve.GroupOrder)
+	// commit to t1
+	tau1 := p.Curve.NewRandomZr(rand)
+	T1 := p.CommitmentGenerators[0].Mul(t1)
+	T1.Add(p.CommitmentGenerators[1].Mul(tau1))
+
+	// compute = \sum y^iU_iV_i
+	t2 := innerProduct(randomLeft, randRightPrime, p.Curve)
+	// commit to t2
+	tau2 := p.Curve.NewRandomZr(rand)
+	T2 := p.CommitmentGenerators[0].Mul(t2)
+	T2.Add(p.CommitmentGenerators[1].Mul(tau2))
+
+	// compute challenge x
+	array = common.GetG1Array([]*math.G1{T1, T2})
+	bytesToHash, err = array.Bytes()
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+	x := p.Curve.HashToZr(bytesToHash)
+
+	// compute vectors left and right against which an IPA will be produced
+	// if p.Value is within the authorized range, then L_iR_i =0 and L_i-R_i-1 = 0
+	// the inner product <left, right> = p.Value*z^2+t1x+t2x^2+f(z, y)
+	// f(z, y) = \sum (z-z^2)*y^i + z^3*2^i
```

Comment:
in the paper `f(z, y) = \sum (z-z^2)*y^i + z^3*2^i` is `f(z, y) = \sum (z-z^2)*y^i - z^3*2^i`, why the `+`?  What am I missing?

Reply:
This is a typo. If you look at the verifier code what's implemented it is what's in the paper.